pkg/gardenlet/bootstrap/certificate/certificate_rotation.go

```diff
@@ -0,0 +1,233 @@
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/retry"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+// certificateWaitTimeout controls the amount of time we wait for the certificate
+// approval in one iteration.
+var certificateWaitTimeout = 15 * time.Minute
+
+// Manager can be used to schedule the certificate rotation for the Gardenlet's Garden cluster client certificate
+type Manager struct {
+	stopCh chan struct{}
+
+	seedName               string
+	clientMap              clientmap.ClientMap
+	gardenClientConnection config.GardenClientConnection
+
+	logger logrus.FieldLogger
+	// Set to time.Now but can be stubbed out for testing
+	now func() time.Time
+}
+
+// NewCertificateRotation creates a certificate manager that can be used to rotate gardenlet's client certificate for the Garden cluster
+func NewCertificateManager(gardenClientConnection config.GardenClientConnection, seedConfig *config.SeedConfig, clientMap clientmap.ClientMap) *Manager {
+	seedName := util.GetSeedName(seedConfig)
+	return &Manager{
+		stopCh:                 make(chan struct{}),
+		seedName:               seedName,
+		clientMap:              clientMap,
+		gardenClientConnection: gardenClientConnection,
+		logger:                 logger.NewFieldLogger(logger.Logger, "certificate-rotation", seedName).Logger,
+		now:                    time.Now,
+	}
+}
+
+// Stop stops the certificate rotation
+func (cr *Manager) Stop() {
+	close(cr.stopCh)
+}
+
+// ScheduleCertificateRotation waits until the currently used Garden cluster client certificate approaches expiration.
+// Then requests a new certificate and stores the kubeconfig in a secret (`gardenClientConnection.kubeconfigSecret`) on the Seed.
+// the argument is context.Cancle function to cancle the context of the Gardenlet used for graceful termination after a successful certificate rotation.
+// when the new gardenlet pod is started, it uses the rotated certificate stored in the secret in the Seed cluster
+func (cr *Manager) ScheduleCertificateRotation(gardenletCancle context.CancelFunc) {
+	go wait.Until(func() {
+		if err := cr.waitForCertificateRotation(); err != nil {
+			cr.logger.Errorf("waiting for the certificate rotation failed: %v", err)
+			return
+		}
+
+		err := retry.Until(context.TODO(), certificateWaitTimeout, func(ctx context.Context) (bool, error) {
+			err := cr.rotateCertificate()
+			if err != nil {
+				cr.logger.Errorf("certificate rotation failed: %v", err)
+				return retry.MinorError(err)
+			}
+			return retry.Ok()
+		})
+		if err != nil {
+			cr.logger.Errorf("Encountered error when trying to rotate certificate: %v", err)
+			return
+		}
+
+		cr.logger.Info("Terminating Gardenlet after successful certificate rotation.")
+		gardenletCancle()
+	}, time.Second, cr.stopCh)
+}
+func (cr *Manager) waitForCertificateRotation() error {
```

Comment:
Why do we have this function? Can't you call `waitForCertificateRotation` directly like here?

Reply:
Its there so that the tests can call these functions without having to build a Manager